File: eel-core/src/main/kotlin/io/eels/component/jdbc/GenericJdbcDialect.kt
package io.eels.component.jdbc

import io.eels.Row
import io.eels.schema.Field
import io.eels.schema.FieldType
import io.eels.schema.Schema
import io.eels.util.Logging
import java.sql.Types

open class GenericJdbcDialect : JdbcDialect, Logging {

  override fun toJdbcType(field: Field): String = when (field.type) {
    FieldType.Long -> "int"
    FieldType.BigInt -> "int"
    FieldType.Int -> "int"
    FieldType.Short -> "smallint"
    FieldType.Boolean -> "BOOLEAN"
    FieldType.String -> {
      if (field.precision.value > 0) "varchar(${field.precision.value})"
      else "varchar(255)"
    }
    else -> {
      logger.warn("Unknown schema type ${field.type}")
      "varchar(255)"
    }
  }

  override fun fromJdbcType(i: Int): FieldType = when (i) {
    Types.BIGINT -> FieldType.Long
    Types.BINARY -> FieldType.Binary
    Types.BIT -> FieldType.Boolean
    Types.BLOB -> FieldType.Binary
    Types.BOOLEAN -> FieldType.Boolean
    Types.CHAR -> FieldType.String
    Types.CLOB -> FieldType.String
    Types.DATALINK -> throw UnsupportedOperationException()
    Types.DATE -> FieldType.Date
    Types.DECIMAL -> FieldType.Decimal
    Types.DISTINCT -> throw UnsupportedOperationException()
    Types.DOUBLE -> FieldType.Double
    Types.FLOAT -> FieldType.Float
    Types.INTEGER -> FieldType.Int
    Types.JAVA_OBJECT -> FieldType.String
    Types.LONGNVARCHAR -> FieldType.String
    Types.LONGVARBINARY -> FieldType.Binary
    Types.LONGVARCHAR -> FieldType.String
    Types.NCHAR -> FieldType.String
    Types.NCLOB -> FieldType.String
    Types.NULL -> FieldType.String
    Types.NUMERIC -> FieldType.Decimal
    Types.NVARCHAR -> FieldType.String
    Types.OTHER -> FieldType.String
    Types.REAL -> FieldType.Double
    Types.REF -> FieldType.String
    Types.ROWID -> FieldType.Long
    Types.SMALLINT -> FieldType.Int
    Types.SQLXML -> FieldType.String
    Types.STRUCT -> FieldType.String
    Types.TIME -> FieldType.Timestamp
    Types.TIMESTAMP -> FieldType.Timestamp
    Types.TINYINT -> FieldType.Int
    Types.VARBINARY -> FieldType.Binary
    Types.VARCHAR -> FieldType.String
    else -> FieldType.String
  }

  override fun create(schema: Schema, table: String): String {
    val columns = schema.fields.map { "${it.name} ${toJdbcType(it)}" }.joinToString(",", "(", ")")
    return "CREATE TABLE $table $columns"
  }

  override fun insertQuery(schema: Schema, table: String): String {
    val columns = schema.fieldNames().joinToString(",")
    val parameters = Array(schema.fields.size, { "?" }).joinToString(",")
    return "INSERT INTO $table ($columns) VALUES ($parameters)"
  }

  override fun insert(row: Row, table: String): String {
    // todo use proper statements
    val columns = row.schema.fieldNames().joinToString(",")
    val values = row.values.joinToString("'", "','", "'")
    return "INSERT INTO $table ($columns) VALUES ($values)"
  }
}

File: eel-core/src/main/scala/io/eels/component/jdbc/GenericJdbcDialect.scala
package io.eels.component.jdbc

import io.eels.Row
import io.eels.schema.Field
import io.eels.schema.FieldType
import io.eels.schema.Schema
import java.sql.Types

import com.sksamuel.exts.Logging

class GenericJdbcDialect extends JdbcDialect with Logging {

  override def toJdbcType(field: Field): String = field.`type` match {
    case FieldType.Long => "int"
    case FieldType.BigInt => "int"
    case FieldType.Int => "int"
    case FieldType.Short => "smallint"
    case FieldType.Boolean => "BOOLEAN"
    case FieldType.String =>
      if (field.precision.value > 0) s"varchar(${field.precision.value})"
      else "varchar(255)"
    case _ =>
      logger.warn(s"Unknown schema type ${field.`type`}")
      "varchar(255)"
    }

  override def fromJdbcType(i: Int): FieldType = i match {
    case Types.BIGINT => FieldType.Long
    case Types.BINARY => FieldType.Binary
    case Types.BIT => FieldType.Boolean
    case Types.BLOB => FieldType.Binary
    case Types.BOOLEAN => FieldType.Boolean
    case Types.CHAR => FieldType.String
    case Types.CLOB => FieldType.String
    case Types.DATALINK => throw new UnsupportedOperationException()
    case Types.DATE => FieldType.Date
    case Types.DECIMAL => FieldType.Decimal
    case Types.DISTINCT => throw new UnsupportedOperationException()
    case Types.DOUBLE => FieldType.Double
    case Types.FLOAT => FieldType.Float
    case Types.INTEGER => FieldType.Int
    case Types.JAVA_OBJECT => FieldType.String
    case Types.LONGNVARCHAR => FieldType.String
    case Types.LONGVARBINARY => FieldType.Binary
    case Types.LONGVARCHAR => FieldType.String
    case Types.NCHAR => FieldType.String
    case Types.NCLOB => FieldType.String
    case Types.NULL => FieldType.String
    case Types.NUMERIC => FieldType.Decimal
    case Types.NVARCHAR => FieldType.String
    case Types.OTHER => FieldType.String
    case Types.REAL => FieldType.Double
    case Types.REF => FieldType.String
    case Types.ROWID => FieldType.Long
    case Types.SMALLINT => FieldType.Int
    case Types.SQLXML => FieldType.String
    case Types.STRUCT => FieldType.String
    case Types.TIME => FieldType.Timestamp
    case Types.TIMESTAMP => FieldType.Timestamp
    case Types.TINYINT => FieldType.Int
    case Types.VARBINARY => FieldType.Binary
    case Types.VARCHAR => FieldType.String
    case _ => FieldType.String
  }

  override def create(schema: Schema, table: String): String = {
    val columns = schema.fields.map { it => s"${it.name} ${toJdbcType(it)}" }.mkString(",", "(", ")")
    s"CREATE TABLE $table $columns"
  }

  override def insertQuery(schema: Schema, table: String): String = {
    val columns = schema.fieldNames().mkString(",")
    val parameters = List.fill(schema.fields.size)("?").mkString(",")
    s"INSERT INTO $table ($columns) VALUES ($parameters)"
  }

  override def insert(row: Row, table: String): String = {
    // todo use proper statements
    val columns = row.schema.fieldNames().mkString(",")
    val values = row.values.mkString("'", "','", "'")
    s"INSERT INTO $table ($columns) VALUES ($values)"
  }
}